src/bokeh/models/widgets/tables.py

```diff
@@ -858,6 +858,50 @@ def __init__(self, *args, **kwargs) -> None:
     The height of each row in pixels.
     """)
 
+    @staticmethod
+    def from_data(data, columns=None, formatters={}, **kwargs) -> DataTable:
+        """ Create a simple table from a pandas dataframe, dictionary or ColumnDataSource.
+
+        Args:
+            data (DataFrame or dict or ColumnDataSource) :
+                The data to create the table from. If the data is a dataframe
+                or dictionary, a ColumnDataSource will be created from it.
+
+            columns (list, optional) :
+                A list of column names to use from the input data.
+                If None, use all columns. (default: None)
+
+            formatters (dict, optional) :
+                A mapping of column names and corresponding Formatters to
+                apply to each column. (default: None)
+
+        Keyword arguments:
+            Any additional keyword arguments will be passed to DataTable
+
+        Returns:
+            DataTable
+
+        Raises:
+            ValueError
+                If the data is not a pandas Dataframe, dictionary or ColumnDataSource
+
+        """
+        import pandas as pd
+
+        if isinstance(data, (pd.DataFrame, dict)):
+            source = ColumnDataSource(data)
+        elif isinstance(data, ColumnDataSource):
+            source = data
+        else:
+            raise ValueError("Data should be a pandas DataFrame, dictionary, or a Bokeh ColumnDataSource.")
+
+        if columns is not None:
+            source.data = {col: source.data[col] for col in columns}
```

Comment:
You are modifying the data source in-place here, effectively making this method take ownership of this data source. If the data source is used for other purpose as well, which is quite common, this will break such code. What's worse, this will happen only for subset of code paths in this method. There are a few ways of dealing with this, depending on what your intentions are for this method. You can either always clone the data source on line 894, or you can clone it only when columns are provided, or you can do nothing and vocally explain to the user that this method modifies the input data source and it's up to the user to clone the data source at the call site if he needs so.

Reply:
This is a good catch, in the case when the user passes in a CDS, we should not modify the CDS> I think it would be better to clone from the start, so that the function always consistently creates a new CDS for all codepaths. If users need to re-use a CDS they can/should just construct the table themselves manually.

Reply:
I'm not sure I completely understand the observations here but I now clone the data only when columns are provided, since otherwise, the data is not modified at all:
```
if columns is not None:
            source = ColumnDataSource(data=dict((col, source.data[col]) for col in columns))
```

Reply:
@Azaya89 I think it would better to clone form the start, and in fact, then we can have an explicit test to maintain that the CDS configured on the table and the CDS that the user passed in are *not* the same. That is a guaranteed way to ensure that we never accidentally modify the user data in the future. 

The cloning @mattpap was referring to is this:
```
In [2]: s = ColumnDataSource(data=dict(a=[1,2,3], b=[4,5,6]))

In [3]: s2 = s.clone()

In [4]: s is s2
Out[4]: False

In [5]: s2.data
Out[5]: {'a': [1, 2, 3], 'b': [4, 5, 6]}
```
If we clone the input CDS at the very start, the the previous code in this spot was fine (because it would be modifying the cloned CDS and that is OK)

Reply:
> ```python
> In [2]: s = ColumnDataSource(data=dict(a=[1,2,3], b=[4,5,6]))
> 
> In [3]: s2 = s.clone()
> 
> In [4]: s is s2 # not the same
> Out[4]: False
> 
> In [5]: s2.data
> Out[5]: {'a': [1, 2, 3], 'b': [4, 5, 6]}
> ```
 For some reason this doesn't work for me. I get a RunTime error when trying this out in Jupyter notebook:

<img width="735" alt="Screenshot 2024-01-12 at 10 40 16 AM" src="https://github.com/bokeh/bokeh/assets/99359668/ba9097ab-8454-4770-bca5-47bbbf6d335e">

`RuntimeError: ColumnDataSource.selected is a readonly property`